### PR TITLE
Add help strings to metrics.

### PR DESCRIPTION
**Pull request by mberhault**
Work towards #9289

I'd like to add a check for non-empty help strings, but I don't want to to had far as a panic.
Would this be a reasonable `make check` step? Grep wouldn't do it though.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14603)

### CODE REVIEW

File: pkg/sql/mem_metrics.go [hunk 1]
```diff
@@ -59,13 +59,25 @@ const log10int64times1000 = 19 * 1000
...
 // MakeMemMetrics instantiates the metric objects for an SQL endpoint.
 func MakeMemMetrics(endpoint string, histogramWindow time.Duration) MemoryMetrics {
-	prefix := "sql.mon." + endpoint
-	MetaMemMaxBytes := metric.Metadata{Name: prefix + ".max"}
-	MetaMemCurBytes := metric.Metadata{Name: prefix + ".cur"}
-	MetaMemMaxTxnBytes := metric.Metadata{Name: prefix + ".txn.max"}
-	MetaMemTxnCurBytes := metric.Metadata{Name: prefix + ".txn.cur"}
-	MetaMemMaxSessionBytes := metric.Metadata{Name: prefix + ".session.max"}
-	MetaMemSessionCurBytes := metric.Metadata{Name: prefix + ".session.cur"}
+	prefix := "sql.mem." + endpoint
```
Comment (mberhault) [inline pkg/sql/mem_metrics.go]: this is the only renamed one. I'd like to clarify `max` and `current`, and probably the caller-specified `endpoint`.
@knz: you can probably shed some light here.

File: pkg/storage/metrics.go [hunk 2]
```diff
@@ -28,288 +28,424 @@ import (
...
+		Help: "Number of failed lease requests"}
+	metaLeaseTransferSuccessCount = metric.Metadata{
+		Name: "leases.transfers.success",
+		Help: "Number of successful lease transfers"}
+	metaLeaseTransferErrorCount = metric.Metadata{
+		Name: "leases.transfers.error",
+		Help: "Number of failed lease transfers"}
+	metaLeaseExpirationCount = metric.Metadata{
+		Name: "leases.expiration",
+		Help: "TODO(marc): fill me"}
```
Comment (mberhault) [inline pkg/storage/metrics.go]: any suggestions? This is very unclear.

File: pkg/sql/mem_metrics.go (hunk 1 above)
Comment (knz) [inline pkg/sql/mem_metrics.go]: "max" is the high water mark of memory allocations for that monitor/category.
"current" is the currently registered allocations for that monitor/category.
There is a bunch of endpoints, they are documented in `sql/mon/mem_usage.go`.

Comment (mberhault) [inline pkg/sql/mem_metrics.go]: wait, the high water mark? why is this a histogram then?

**Comment by knz** [comment]
The high water mark is measured per connection. The histogram collects the HWMs over all connections.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: pkg/storage/metrics.go (hunk 2 above)
Comment (mberhault) [inline pkg/storage/metrics.go]: @bdarnell: ideas?

File: pkg/sql/mem_metrics.go (hunk 1 above)
Comment (mberhault) [inline pkg/sql/mem_metrics.go]: ok, then I think the current help strings I have are fine.

File: pkg/storage/metrics.go (hunk 2 above)
Comment (bdarnell) [inline pkg/storage/metrics.go]: It's "Number of replicas on this store with {expiration,epoch}-based leases". These are not very useful metrics - a small number of system ranges use expiration-based leases; all others are epoch-based. The two sum to the number of replicas on the node. I would just get rid of them.

Comment (a-robinson) [inline pkg/storage/metrics.go]: I've found it mildly useful at times to see that the number of expiration-based leases in a cluster dropped down from its steady state, indicating a system range is unavailable. It's not useful for much else, though, and other metrics could also cover that case.

Comment (mberhault) [inline pkg/storage/metrics.go]: Thanks, I've updated the description. Let's keep those for now.

**Comment by a-robinson** [comment]
:lgtm:

---

Reviewed 9 of 12 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/kv/txn_coord_sender.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTlA-PTHA64uSxgqg1:-KhTlA-PTHA64uSxgqg2:b-i9x5h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/kv/txn_coord_sender.go#L126)):*
> ```Go
> 	metaCommits1PCRates = metric.Metadata{
> 		Name: "txn.commits1PC",
> 		Help: "Number of committed one-piece KV transactions"}
> ```

s/one-piece/one-phase/

---

*[pkg/server/node.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTlMPm1OdohjgV4q9k:-KhTlMPm1OdohjgV4q9l:b-ujzxck) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/server/node.go#L69)):*
> ```Go
> 	metaExecLatency = metric.Metadata{
> 		Name: "exec.latency",
> 		Help: "Latency of batch calls executed on this node"}
> ```

I might try to work "kv" in to these three help messages to make people less likely to confuse them with sql metrics. Something like "Latency of batch KV requests executed locally on this node"

---

*[pkg/sql/executor.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTmYN0bGOp4qrtxVl4:-KhTmYN1qkh_ErCo0zwR:b2frvag) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/sql/executor.go#L68)):*
> ```Go
> 	MetaTxnBegin = metric.Metadata{
> 		Name: "sql.txn.begin.count",
> 		Help: "Number of begun sql transactions"}
> ```

Some of these descriptions aren't quite precise enough. IIUC, the metrics don't measure all begun SQL transactions (or all committed/aborted transactions), but the number of times that a `BEGIN` statement was executed. It's a minor difference, but seems important, particularly for the number of committed transactions. Implicit single-statement transactions get committed, but don't get counted in the "committed" counter (or counted as begun or aborted, etc.).

---

*[pkg/sql/executor.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTn4WB-HZPW9oQLQPW:-KhTn4WCt3yJzDkbtdU4:b-8zozjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/sql/executor.go#L80)):*
> ```Go
> 	MetaSelect = metric.Metadata{
> 		Name: "sql.select.count",
> 		Help: "Number of sql select calls"}
> ```

s/calls/statements/ for all of these help messages

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-:-KhTo11bEN1RgeyL9ArR:bnfp4nl)*

**Comment by mberhault** [comment]
Review status: 9 of 12 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTlA-PTHA64uSxgqg1:-KhXPiSqhJ_FGM86R9lt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/kv/txn_coord_sender.go#L126)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/one-piece/one-phase/
</blockquote></details>

Done.

---

*[pkg/server/node.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTlMPm1OdohjgV4q9k:-KhXPs7t5RtfuF70PR7k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/server/node.go#L69)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I might try to work "kv" in to these three help messages to make people less likely to confuse them with sql metrics. Something like "Latency of batch KV requests executed locally on this node"
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTmYN0bGOp4qrtxVl4:-KhXQUIuSRummR_QYUE5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/sql/executor.go#L68)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Some of these descriptions aren't quite precise enough. IIUC, the metrics don't measure all begun SQL transactions (or all committed/aborted transactions), but the number of times that a `BEGIN` statement was executed. It's a minor difference, but seems important, particularly for the number of committed transactions. Implicit single-statement transactions get committed, but don't get counted in the "committed" counter (or counted as begun or aborted, etc.).
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 80 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603#-KhTn4WB-HZPW9oQLQPW:-KhXQVJoyS0ul3vM4FuP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2453794940e35afebbe4c547f61b2e99b7a3715/pkg/sql/executor.go#L80)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/calls/statements/ for all of these help messages
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603)*

**Comment by a-robinson** [comment]
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14603)*

**Pull request merged**